src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs

```diff
@@ -11,28 +11,53 @@
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Formatting.Rules;
+using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Shared.Utilities;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.Formatting
 {
-    internal class IndentBlockFormattingRule : BaseFormattingRule
+    internal sealed class IndentBlockFormattingRule : BaseFormattingRule
     {
         internal const string Name = "CSharp IndentBlock Formatting Rule";
 
-        public override void AddIndentBlockOperations(List<IndentBlockOperation> list, SyntaxNode node, AnalyzerConfigOptions options, in NextIndentBlockOperationAction nextOperation)
+        private readonly CachedOptions _options;
+
+        public IndentBlockFormattingRule()
+            : this(new CachedOptions(null))
+        {
+        }
+
+        private IndentBlockFormattingRule(CachedOptions options)
+        {
+            _options = options;
+        }
+
+        public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options)
+        {
+            var cachedOptions = new CachedOptions(options);
+
+            if (cachedOptions == _options)
+            {
+                return this;
+            }
```

Comment:
so we basically reuse if user hasn't change from the defaults.  but we always allocate if they have?  woudl def like an approach where we can get the benefits as well for people not on the defaults.

Reply:
In my measurements it didn't matter. <0.1%.

Reply:
if that's the case, why do we even need the "return this" optimization?

Reply:
It's the standard pattern for 'with' methods.

Reply:
Fair enough. :)